### PR TITLE
Mount options for disk device mounts #6225

### PR DESCRIPTION
**Pull request by splett2**
This PR adds support for passing in raw mount option string for disk device mounts, e.g: 

lxc config device add first vfat1 disk source=/dev/loop9 path=/mnt/test raw.mount.options=gid=123,uid=456

### CODE REVIEW

File: lxd/device/device_utils_disk.go [hunk 1]
```diff
@@ -95,22 +95,22 @@ func DiskMount(srcPath string, dstPath string, readonly bool, recursive bool, pr
...
+	err = unix.Mount(srcPath, dstPath, fstype, uintptr(flags), rawMountOptions)
 	if err != nil {
 		return fmt.Errorf("Unable to mount %s at %s: %s", srcPath, dstPath, err)
 	}
 
 	// Remount bind mounts in readonly mode if requested
 	if readonly == true && flags&unix.MS_BIND == unix.MS_BIND {
 		flags = unix.MS_RDONLY | unix.MS_BIND | unix.MS_REMOUNT
-		err = unix.Mount("", dstPath, fstype, uintptr(flags), "")
+		err = unix.Mount("", dstPath, fstype, uintptr(flags), rawMountOptions)
```
Comment (stgraber) [inline lxd/device/device_utils_disk.go]: You do not need to pass it to that one.

File: lxd/device/device_utils_disk.go [hunk 2]
```diff
@@ -95,22 +95,22 @@ func DiskMount(srcPath string, dstPath string, readonly bool, recursive bool, pr
...
 		if err != nil {
 			return fmt.Errorf("Unable to mount %s in readonly mode: %s", dstPath, err)
 		}
 	}
 
 	flags = unix.MS_REC | unix.MS_SLAVE
-	err = unix.Mount("", dstPath, "", uintptr(flags), "")
+	err = unix.Mount("", dstPath, "", uintptr(flags), rawMountOptions)
```
Comment (stgraber) [inline lxd/device/device_utils_disk.go]: Nor to that one.

File: doc/containers.md [hunk 3]
```diff
@@ -495,7 +495,7 @@ recursive        | boolean   | false             | no        | Whether or not to
 pool             | string    | -                 | no        | The storage pool the disk device belongs to. This is only applicable for storage volumes managed by LXD.
 propagation      | string    | -                 | no        | Controls how a bind-mount is shared between the container and the host. (Can be one of `private`, the default, or `shared`, `slave`, `unbindable`,  `rshared`, `rslave`, `runbindable`,  `rprivate`. Please see the Linux Kernel [shared subtree](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt) documentation for a full explanation)
 shift            | boolean   | false             | no        | Setup a shifting overlay to translate the source uid/gid to match the container
-raw.mount.options| string    | -      			 | no        | Filesystem specific mount options 
+raw.mount.options| string    | -                 | no        | Filesystem specific mount options
```
Comment (stgraber) [inline doc/containers.md]: This alignment fix should be applied to the `doc` commit instead.

**Comment by stgraber** [comment]
@splett2 @Lilyjjo this looks very good, thanks, just a few minor nits that I commented on in-line above.

And I also noticed that the `Signed-off-by` line looks slightly incorrect.
I see it as `Signed-off-by: Lillian J. Johnson lillianjanjohnson@gmail.com` instead of `Signed-off-by: Lillian J. Johnson <lillianjanjohnson@gmail.com>`.

One more thing about sign-off, it's usually good practice to have one such Signed-off-by line per author when a commit was authored or significantly altered by multiple people, effectively keeping track of everyone's contribution to the work.


It's something you'll see much more commonly in projects like the Linux kernel where a feature is often developed by 2-3 different people, and reviewd/debated on public mailing-lists, resulting in multiple lines like:
 - Signed-off-by
 - Suggested-by
 - Acked-by
 - Reviewed-by
 - ...

Since we do our reviews on Github, we don't use `Acked-by`, instead having the Github merge commit act as that one. But authors can still make use of the others. You'd typically use `Suggested-by` if doing work that's not already tied to a Github issue to indicate that someone else pointed out the issue or missing feature to you. `Reviewed-by` would be if you had someone who's not on Github or didn't comment in the PR do a review for you ahead of time.

**Comment by splett2** [comment]
Thanks, we fixed up the commits.

**Comment by stgraber** [comment]
Looks good to me, let's see if Jenkins agrees.

**Comment by stgraber** [comment]
jenkins: test this please

File: test/suites/container_devices_disk.sh [hunk 4]
```diff
@@ -59,3 +60,28 @@ test_container_devices_disk_shift() {
   lxc storage volume delete "${POOL}" foo-shift
   lxc stop foo -f
 }
+
+test_container_devices_raw_mount_options() {
+  configure_loop_device loop_file_1 loop_device_1
+  mkfs.vfat "${loop_device_1}"
```
Comment (stgraber) [inline test/suites/container_devices_disk.sh]: Needs a `# shellcheck disable=SC2154` comment above is so shellcheck doesn't freak out.

File: test/suites/container_devices_disk.sh [hunk 5]
```diff
@@ -59,3 +60,28 @@ test_container_devices_disk_shift() {
...
+
+  lxc stop foo-priv -f
+  lxc config device add foo-priv loop_raw_mount_options disk source="${loop_device_1}" path=/mnt raw.mount.options=uid=123,gid=456
+  lxc start foo-priv
+
+  [ "$(lxc exec foo -- stat /mnt -c '%u:%g')" = "123:456" ] || false
+  lxc config device remove foo-priv loop_raw_mount_options
+
+  lxc delete -f foo-priv
+  deconfigure_loop_device "${loop_file_1}" "${loop_device_1}"
```
Comment (stgraber) [inline test/suites/container_devices_disk.sh]: Needs a `# shellcheck disable=SC2154` comment above is so shellcheck doesn't freak out.

File: test/suites/container_devices_disk.sh [hunk 6]
```diff
@@ -59,3 +60,28 @@ test_container_devices_disk_shift() {
...
+  
+  lxc config device add foo-priv loop_raw_mount_options disk source="${loop_device_1}" path=/mnt raw.mount.options=uid=123,gid=456
+  [ "$(lxc exec foo -- stat /mnt -c '%u:%g')" = "123:456" ] || false
+  lxc config device remove foo-priv loop_raw_mount_options
+
+  lxc stop foo-priv -f
+  lxc config device add foo-priv loop_raw_mount_options disk source="${loop_device_1}" path=/mnt raw.mount.options=uid=123,gid=456
+  lxc start foo-priv
+
+  [ "$(lxc exec foo -- stat /mnt -c '%u:%g')" = "123:456" ] || false
```
Comment (stgraber) [inline test/suites/container_devices_disk.sh]: Wrong name, container is `foo-priv` not `foo`.

File: test/suites/container_devices_disk.sh [hunk 7]
```diff
@@ -59,3 +60,28 @@ test_container_devices_disk_shift() {
...
 }
+
+test_container_devices_raw_mount_options() {
+  configure_loop_device loop_file_1 loop_device_1
+  mkfs.vfat "${loop_device_1}"
+
+  lxc launch testimage foo-priv -c security.privileged=true
+
+  lxc config device add foo-priv loop_raw_mount_options disk source="${loop_device_1}" path=/mnt
+  [ "$(lxc exec foo -- stat /mnt -c '%u:%g')" = "0:0" ] || false
```
Comment (stgraber) [inline test/suites/container_devices_disk.sh]: Wrong name, container is `foo-priv` not `foo`.

File: test/suites/container_devices_disk.sh [hunk 8]
```diff
@@ -59,3 +60,28 @@ test_container_devices_disk_shift() {
...
+  lxc config device remove foo-priv loop_raw_mount_options
+  
+  lxc config device add foo-priv loop_raw_mount_options disk source="${loop_device_1}" path=/mnt raw.mount.options=uid=123,gid=456
+  [ "$(lxc exec foo -- stat /mnt -c '%u:%g')" = "123:456" ] || false
```
Comment (stgraber) [inline test/suites/container_devices_disk.sh]: Wrong name, container is `foo-priv` not `foo`.

**Comment by stgraber** [comment]
@splett2 almost there, just a few issues with the test and we should be good.

**Comment by splett2** [comment]
Thanks, added in the fix. We'll try to get the test suite working locally so this won't be an issue next time.

**Comment by stgraber** [comment]
jenkins: test this please

File: test/suites/container_devices_disk.sh (hunk 4 above)
Comment (stgraber) [reply test/suites/container_devices_disk.sh]: The comment must go above the mkfs.vfat line, not the configure_loop_device.

As far as I can tell, that's the only issue left before we can merge this.

Comment (splett2) [reply test/suites/container_devices_disk.sh]: Updated

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
```
ERROR: trailing whitespace in script: ./test/suites/container_devices_disk.sh:  
```

File: test/suites/container_devices_disk.sh [hunk 9]
```diff
@@ -59,3 +60,30 @@ test_container_devices_disk_shift() {
...
+  configure_loop_device loop_file_1 loop_device_1
+  # shellcheck disable=SC2154
+  mkfs.vfat "${loop_device_1}"
+
+  lxc launch testimage foo-priv -c security.privileged=true
+
+  lxc config device add foo-priv loop_raw_mount_options disk source="${loop_device_1}" path=/mnt
+  [ "$(lxc exec foo-priv -- stat /mnt -c '%u:%g')" = "0:0" ] || false
+  lxc config device remove foo-priv loop_raw_mount_options
+
```
Comment (stgraber) [inline test/suites/container_devices_disk.sh]: That's where the whitespace is hiding

Comment (splett2) [reply test/suites/container_devices_disk.sh]: Thanks so much for your patience through all of this!

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
More static analysis failures I'm afraid, this time it's alignment issues.

You can fix those by running 'go fmt ./...'

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
And it's all green!

**Comment by stgraber** [comment]
Thanks for this work!

**Pull request merged**